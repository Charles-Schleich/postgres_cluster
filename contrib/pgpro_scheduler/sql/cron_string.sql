create user __temp_robot;
create user __temp_root WITH  SUPERUSER;

SET SESSION AUTHORIZATION __temp_root;

select schedule.create_job(
    '{
        "name": "Test @reboot",
        "cron": "@reboot",
        "command": "show all",
        "run_as": "__temp_robot"
    }'
);

select schedule.create_job(
    '{
        "name": "Test 1",
        "cron": "* * * * *",
        "command": "select ''this is every minute job''",
        "run_as": "__temp_robot",
        "last_start_available": "2 hours"
    }'
);

select schedule.create_job(
    '{
        "name": "Test 2 4/4 2/4 * * *",
        "cron": "4/4 2/4 * * *",
        "command": "select pg_sleep(10)",
        "run_as": "__temp_robot"
    }'
);

select schedule.create_job(
	'{
		"name": "Test 3",
		"cron": "23 1 * * THU,SUN",
		"command": "select ''ok'' as ok"
	}'
);

select id, node, name, rule, do_sql, same_transaction, postpone, retry from schedule.cron order by id;

RESET SESSION AUTHORIZATION;
drop user __temp_root;
drop user __temp_robot;
