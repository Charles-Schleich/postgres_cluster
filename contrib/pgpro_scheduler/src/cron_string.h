#ifndef CRON_STRING_PARSE_H
#define CRON_STRING_PARSE_H

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>

extern int cps_error;

#define CE_MINUTES_LEN	60
#define CE_HOURS_LEN	24
#define CE_DAYS_LEN		31
#define CE_MONTHS_LEN	12
#define CE_DOWS_LEN		7

typedef struct {
	int start;
	int end;
	char *subst;
	int subst_len;
} match_ent_t;

typedef struct {
	int mins_len;
	int *mins;
	
	int hour_len;
	int *hour;

	int day_len;
	int *day;

	int month_len;
	int *month;

	int wdays_len;
	int *wdays;

	unsigned char onstart;
} cron_ent_t;

void destroyCronEnt(cron_ent_t *);
void _cps_set_error(int num, const char *message, ...) 
#ifdef __GNUC__
__attribute__ ((format (gnu_printf, 2, 3)))
#endif
;;
char *get_cps_error(void);
int _cps_string_has(char *str, char c);
char *_cps_append_string(char *str, char *to_add);
char *_cps_make_array(int *src, int len);
int _cps_dig_digit(char *str);
void _cps_clean_charchar(char **data, int len);
char **_cps_split_by(char sep, char *src, int *N);
int *_cps_parse_range(char *src, int start, int end, int *len);
char *_cps_subst_str(char *str, char **subst_array, int subst_len, int step);
cron_ent_t *parse_crontab(char *cron_str);
char *parse_crontab_to_json_text(char *cron_str);

#endif
