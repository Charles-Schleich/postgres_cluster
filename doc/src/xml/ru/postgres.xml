<?xml version="1.0" encoding="utf-8"?>
<!-- doc/src/xml/postgres.xml -->

<!DOCTYPE book PUBLIC "-//OASIS//DTD DocBook XML V4.2//EN" "http://www.oasis-open.org/docbook/xml/4.2/docbookx.dtd" [

<!ENTITY % version SYSTEM "version.ent">
%version;
<!ENTITY % filelist SYSTEM "filelist.xml">
%filelist;
<!ENTITY % isoent SYSTEM "http://www.w3.org/2003/entities/2007/isopub.ent"> %isoent;
<!ENTITY % xhtml1-lat1 SYSTEM "http://www.w3.org/2003/entities/2007/xhtml1-lat1.ent"> %xhtml1-lat1;
<!ENTITY % isolat2 SYSTEM "http://www.w3.org/2003/entities/2007/isolat2.ent"> %isolat2;
<!ENTITY % isogrk3 SYSTEM "http://www.w3.org/2003/entities/2007/isogrk3.ent"> %isogrk3;

<!ENTITY reference  SYSTEM "reference.xml">

]>
<book id="postgres" lang="ru">
 <title>Документация к &productname; &version;</title>

 <bookinfo>
  <corpauthor>The PostgreSQL Global Development Group</corpauthor>   
  <corpauthor>ООО «Постгрес Профессиональный»</corpauthor>
  <productname>&productname;</productname>
  <productnumber>&version;</productnumber>
  &legal;
  <othercredit role="translator">
    <affiliation>
      <orgname>Компания «Постгрес Профессиональный»</orgname>
    </affiliation>
    <contrib>Перевод на русский язык, 2015-2017гг.</contrib>
  </othercredit>
  <releaseinfo>
    Официальная страница русскоязычной документации:
    <ulink url="https://postgrespro.ru/docs/">https://postgrespro.ru/docs/</ulink>.
  </releaseinfo>
 </bookinfo>

 &intro;

 <part id="tutorial">
  <title>Введение</title>

  <partintro>
   <para>Добро пожаловать на встречу с <productname>&productname;</productname>. В следующих главах вы сможете получить общее представление о <productname>&productname;</productname>, реляционных базах данных и языке SQL, если вы ещё не знакомы с этими темами. Этот материал будет понятен даже тем, кто обладает только общими знаниями о компьютерах. Ни опыт программирования, ни навыки использования Unix-систем не требуются. Основная цель этой части &mdash; познакомить вас на практике с ключевыми аспектами системы <productname>&productname;</productname>, поэтому затрагиваемые в ней темы не будут рассматриваться максимально глубоко и полно.</para>

   <para>Прочитав это введение, вы, возможно, захотите перейти к <xref remap="3" linkend="sql"/>, чтобы получить более формализованное знание языка SQL, или к <xref remap="3" linkend="client-interfaces"/>, посвящённой разработке приложений для <productname class="trade">PostgreSQL</productname>. Для изучения SQL можно воспользоваться демонстрационной базой данных, описанной в <xref remap="6" linkend="demodb-bookings"/>. Тем же, кто устанавливает и администрирует сервер самостоятельно, следует также прочитать <xref remap="4" linkend="admin"/>.</para>
  </partintro>

  &start;
  &query;
  &advanced;

 </part>

 <part id="sql">
  <title>Язык SQL</title>

  <partintro>
   <para>В этой части книги описывается использование языка <acronym>SQL</acronym> в <productname>&productname;</productname>. Мы начнём с описания общего синтаксиса <acronym>SQL</acronym>, затем расскажем, как создавать структуры для хранения данных, как наполнять базу данных и как выполнять запросы к ней. В продолжении будут перечислены существующие типы данных и функции, применяемые с командами <acronym>SQL</acronym>. И наконец, закончится эта часть рассмотрением важных аспектов настройки базы данных для оптимальной производительности.</para>

   <para>Материал этой части упорядочен так, чтобы новичок мог прочитать её от начала до конца и полностью понять все темы, не забегая вперёд. При этом главы сделаны самодостаточными, так что опытные пользователи могут читать главы по отдельности. Информация в этой части книги представлена в повествовательном стиле и разделена по темам. Если же вас интересует формальное и полное описание определённой команды, см. <xref remap="4" linkend="reference"/>.</para>

   <para>Читатели этой части книги должны уже знать, как подключаться к базе данных <productname>&productname;</productname> и выполнять команды <acronym>SQL</acronym>. Если вы ещё не знаете этого, рекомендуется сначала прочитать <xref remap="4" linkend="tutorial"/>. Команды <acronym>SQL</acronym> обычно вводятся в <application>psql</application> &mdash; интерактивном терминальном приложении <productname>&productname;</productname>, но можно воспользоваться и другими программами с подобными функциями.</para>
  </partintro>

  &syntax;
  &ddl;
  &dml;
  &queries;
  &datatype;
  &func;
  &typeconv;
  &indices;
  &textsearch;
  &mvcc;
  &perform;
  &parallel;
  &atx;

 </part>

 <part id="admin">
  <title>Администрирование сервера</title>

  <partintro>
   <para>В этой части документации освещаются темы, представляющие интерес для администратора баз данных <productname>&productname;</productname>. В частности, здесь рассматривается установка программного обеспечения, установка и настройка сервера, управление пользователями и базами данных, а также задачи обслуживания. С этими темами следует ознакомиться всем, кто эксплуатирует сервер <productname>&productname;</productname> (даже для личных целей, а тем более в производственной среде).</para>

   <para>Материал этой части даётся примерно в том порядке, в каком его следует читать начинающему пользователю. При этом её главы самодостаточны и при желании могут быть прочитаны по отдельности. Информация в этой части книги представлена в повествовательном стиле и разделена по темам. Если же вас интересует формальное и полное описание определённой команды, см. <xref remap="4" linkend="reference"/>.</para>

   <para>Первые несколько глав написаны так, чтобы их можно было понять без предварительных знаний, так что начинающие пользователи, которым нужно установить свой собственный сервер, могут начать свой путь с них. Остальные главы части посвящены настройке сервера и управлению им; в этом материале подразумевается, что читатель знаком с основными принципами использования СУБД <productname>&productname;</productname>. За дополнительной информацией мы рекомендуем читателям обратиться к <xref remap="3" linkend="tutorial"/> и <xref remap="3" linkend="sql"/>.</para>
  </partintro>

  &installation-bin;
  &runtime;
  &config;
  &client-auth;
  &user-manag;
  &manage-ag;
  &charset;
  &maintenance;
  &backup;
  &high-availability;
  &recovery-config;
  &monitoring;
  &diskusage;
  &wal;
  &regress;
  &cfs;

 </part>

 <part id="client-interfaces">
  <title>Клиентские интерфейсы</title>

  <partintro>
   <para>В этой части документации описываются клиентские программные интерфейсы, включённые в дистрибутив <productname>&productname;</productname>. Все включённые в неё главы можно читать по отдельности. Заметьте, что существует множество других программных интерфейсов, со своей документацией (некоторые наиболее популярные перечислены в <xref remap="6" linkend="external-projects"/>). Для изучения этой части нужно уметь работать с базой данных, используя команды <acronym>SQL</acronym> (см. <xref remap="4" linkend="sql"/>), и, конечно же, знать язык программирования, на который ориентирован определённый интерфейс.</para>
  </partintro>

  &libpq;
  &lobj;
  &ecpg;
  &infoschema;

 </part>

 <part id="server-programming">
  <title>Серверное программирование</title>

  <partintro>
   <para>Эта часть документации посвящена расширению функциональности сервера путём реализации собственных функций, типов данных, триггеров и т. д. Это довольно сложные темы, для освоения которых рекомендуется предварительно изучить и понять всю остальную документацию для пользователей <productname>&productname;</productname>. В последних главах этой части описываются языки программирования на стороне сервера, поддерживаемые дистрибутивом <productname>&productname;</productname>, и рассматриваются общие вопросы, связанные с программированием на стороне сервера. Но прежде чем погружаться в этот материал, важно изучить предыдущие разделы <xref remap="2" linkend="extend"/> (где освещаются функции).</para>
  </partintro>

  &extend;
  &trigger;
  &event-trigger;
  &rules;

  &xplang;
  &plsql;
  &pltcl;
  &plperl;
  &plpython;

  &spi;
  &bgworker;
  &logicaldecoding;
  &replication-origins;

 </part>

 &reference;

 <part id="internals">
  <title>Внутреннее устройство</title>

  <partintro>
   <para>Содержит разнообразную информацию, полезную для разработчиков <productname>&productname;</productname>.</para>
  </partintro>

  &arch-dev;
  &catalogs;
  &protocol;
  &plhandler;
  &fdwhandler;
  &tablesample-method;
  &custom-scan;
  &geqo;
  &indexam;
  &generic-wal;
  &gist;
  &spgist;
  &gin;
  &brin;
  &storage;
  &bki;
  &planstats;

 </part>

 <part id="appendixes">
  <title>Приложения</title>

  &errcodes;
  &datetime;
  &keywords;
  &features;
  &release;
  &contrib;
  &external-projects;
  &config-one-c;
  &demodb-bookings;
  &acronyms;

 </part>

 &biblio;
 <index id="bookindex"/>

</book>
