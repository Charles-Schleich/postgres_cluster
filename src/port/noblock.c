/*-------------------------------------------------------------------------
 *
 * noblock.c
 *	  set a file descriptor as non-blocking
 *
 * Portions Copyright (c) 1996-2016, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * IDENTIFICATION
 *	  src/port/noblock.c
 *
 *-------------------------------------------------------------------------
 */

#include "c.h"

#include <fcntl.h>
#ifdef WITH_RSOCKET
#include <rdma/rsocket.h>
#endif


bool
pg_set_noblock(pgsocket sock, bool isRsocket)
{
#if !defined(WIN32)
#ifdef WITH_RSOCKET
	if (isRsocket)
		return (rfcntl(sock, F_SETFL, O_NONBLOCK) != -1);
	else
#endif
		return (fcntl(sock, F_SETFL, O_NONBLOCK) != -1);
#else
	unsigned long ioctlsocket_ret = 1;

	/* Returns non-0 on failure, while fcntl() returns -1 on failure */
	return (ioctlsocket(sock, FIONBIO, &ioctlsocket_ret) == 0);
#endif
}


bool
pg_set_block(pgsocket sock, bool isRsocket)
{
#if !defined(WIN32)
	int			flags;

#ifdef WITH_RSOCKET
	if (isRsocket)
	{
		flags = rfcntl(sock, F_GETFL);
		if (flags < 0 || rfcntl(sock, F_SETFL, (long) (flags & ~O_NONBLOCK)))
			return false;
	}
	else
#endif
	{
		flags = fcntl(sock, F_GETFL);
		if (flags < 0 || fcntl(sock, F_SETFL, (long) (flags & ~O_NONBLOCK)))
			return false;
	}
	return true;
#else
	unsigned long ioctlsocket_ret = 0;

	/* Returns non-0 on failure, while fcntl() returns -1 on failure */
	return (ioctlsocket(sock, FIONBIO, &ioctlsocket_ret) == 0);
#endif
}
