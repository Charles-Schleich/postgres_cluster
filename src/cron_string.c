#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include "cron_string.h"
#include "postgres.h"
#include "c.h"
#include "port.h"


char *cps_month_subst_data[12] = { "JAN", "FEB", "MAR", "APR", "MAY", "JUN", "JUL", "AUG", "SEP", "OCT", "NOV", "DEC" };
char *cps_wday_subst_data[7] = { "SUN", "MON", "TUE", "WED", "THU", "FRI", "SAT" };

int cps_error = 0;
static char cps_error_buffer[512];

void _cps_set_error(int num, const char *message, ...)
{
	va_list arglist;

	cps_error = num;
	va_start(arglist, message);
	pvsnprintf(cps_error_buffer, 512, message, arglist);
	va_end(arglist);
}

char *get_cps_error()
{
	if(cps_error)
	{
		return cps_error_buffer;
	}
	return NULL;
}

int _cps_string_has(char *str, char c)
{
	int i;
	int n = strlen(str);

	for(i=0; i < n; i++)
	{
		if(c == str[i]) return 1;
	}
	return 0;
}

char *_cps_append_string(char *str, char *to_add)
{
	int end = str ? strlen(str): 0;
	int len = strlen(to_add);
	
	str = str? repalloc(str, end + len + 1): palloc(end + len + 1);
	memcpy(str+end, to_add, len);
	str[end+len] = 0;

	return str;
}

char *_cps_make_array(int *src, int len)
{
	char *out = NULL;
	int i = 0;
	char buff[100];

	out = _cps_append_string(out, "[");
	for(i=0; i<len; i++)
	{
		sprintf(buff, "%d,", src[i]); 
		if(len == (i+1)) buff[strlen(buff)-1] = 0;
		out = _cps_append_string(out, buff);
	}
	return _cps_append_string(out, "]");
}

int _cps_dig_digit(char *str)
{
	int len = strlen(str);
	char *endptr = NULL;
	int res;
	
	if(len == 1 && str[0] == '*')
	{
		return -1;
	}
	res = (int)(strtol(str, &endptr, 10));
	if(endptr == str)
	{
		cps_error = 1;
		_cps_set_error(44, "the %s should be digit", str);
		return -2;
	}
	return res;
}

void _cps_clean_charchar(char **data, int len)
{
	int i;

	for(i=0; i < len; i++)
	{
		pfree(data[i]);
	}
	pfree(data);
}

char **_cps_split_by(char sep, char *src, int *N)
{
	int len = strlen(src);
	char **res = NULL;
	int i, n = 1, ri = 0, ti = 0;
	char tmp[1024];
	
	for(i=0; i<len; i++)
	{
		if(src[i] == sep) n++;
	}
	res = palloc(sizeof(char *)*n);
	*N = n;
	
	for(i=0; i < len; i++)
	{
		if(src[i] == sep)
		{
			tmp[ti++] = 0;
			res[ri] = (char *)palloc(sizeof(char)*(ti));
			memcpy(res[ri++], tmp, ti);
			ti = 0;
		}
		else
		{
			if(ti < 1024) tmp[ti++] = src[i];
		}
	}
	tmp[ti++] = 0;
	res[ri] = palloc(sizeof(char)*(ti+1));
	memcpy(res[ri], tmp, ti+1);

	return res;
}

int *_cps_parse_range(char *src, int start, int end, int *len)
{
	char *values = NULL;
	int hasStar = 0;
	int range_len = end-start+1;
	int i, a, b, cur;
	int ni = 0;
	char **parts;
	int e_len = 0, nparts = 0;
	char **elements;
	int isError = 0;
	int *ptr = NULL;

	*len = 0;

	values = palloc(sizeof(char)*range_len);
	memset(values, 0, sizeof(char)*range_len);

	parts = _cps_split_by(',', src, &nparts);

	for(i=0; i < nparts; i++)
	{
		if(strlen(parts[i]) == 1 && parts[i][0] == '*')
		{
			hasStar = 1;
			break;
		}
	}
	if(hasStar)
	{
		memset(values, 1, sizeof(char)*range_len);
	}
	else
	{
		for(i=0; i < nparts; i++)
		{
			if(_cps_string_has(parts[i], '-'))
			{
				elements = _cps_split_by('-', parts[i], &e_len);
				if(e_len == 2)
				{
					a = _cps_dig_digit(elements[0]);
					if(a < 0)
					{
						_cps_clean_charchar(elements, e_len);
						isError = 1;
						if(a == -1) _cps_set_error(1, "star is not allowed in range");
						break;
					}
					b = _cps_dig_digit(elements[1]);
					if(b < 0)
					{
						_cps_clean_charchar(elements, e_len);
						isError = 2;
						if(b == -1) _cps_set_error(2, "star is not allowed in range");
						break;
					}
					if(b < a)
					{
						_cps_clean_charchar(elements, e_len);
						isError = 3;
						_cps_set_error(2, "bad  range: %d - %d", a, b);
						break;
					}
					for(i = 0; i < range_len; i++)
					{
						if(i >= (a-start) && i <= (b-start)) values[i] = 1; 
					}
					_cps_clean_charchar(elements, e_len);
				}
				else
				{
					_cps_clean_charchar(elements, e_len);
					isError = 4;
					_cps_set_error(4, "range must be of 2 elements: %s", parts[i]);
					break;
				}
			}
			else if(_cps_string_has(parts[i], '/'))
			{
				elements = _cps_split_by('/', parts[i], &e_len);
				if(e_len == 2)
				{
					a = _cps_dig_digit(elements[0]);
					if(a < -1)
					{
						isError = 5;
						_cps_clean_charchar(elements, e_len);
						break;
					}
					if(a == -1) a = start;
					b = _cps_dig_digit(elements[1]);
					if(b <= 0)
					{
						isError = 6;
						_cps_clean_charchar(elements, e_len);
						if(b == -1) _cps_set_error(6, "star is not allowed as period: %s", parts[i]);
						if(b == 0) _cps_set_error(6, "period must be greater than zero: %s", parts[i]);
						break;
					}
					cur = a -start;
					while(cur <= (end+start))
					{
						values[cur] = 1;
						cur += b;
					}
				}
				else
				{
					isError = 10;
					_cps_set_error(10, "wrong period format: %s", parts[i]);
					_cps_clean_charchar(elements, e_len);
					break;
				}
				_cps_clean_charchar(elements, e_len);
			}
			else
			{
				a = _cps_dig_digit(parts[i]);
				if(a < start)
				{
					isError = 11;
					if(a > -1) _cps_set_error(11, "%d is out of range", a);
					break;
				}
				if(a > end)
				{
					isError = 12;
					_cps_set_error(12, "%d is out of range", a);
					break;
				}
				values[a-start] = 1;
			}
		} 
	}
	
	if(isError == 0)
	{
		for(i =0; i < range_len; i++)
		{
			if(values[i]) (*len)++;
		}
			
		if(*len > 0)
		{
			ptr = palloc(sizeof(int)*(*len));
		}
		else
		{
			ptr = NULL;
		}
		
		for(i = 0; i < range_len; i++)
		{
			if(values[i]) ptr[ni++] = i+start;
		}
	}
	else
	{
		ptr = NULL;
		ni = 0;
	}
	pfree(values);
	_cps_clean_charchar(parts, nparts);

	return ptr;
}

char *_cps_subst_str(char *str, char **subst_array, int subst_len, int step)
{
	int len = strlen(str);
	char *new_str =  NULL;
	int *candidats = (int *)palloc(sizeof(int)*subst_len);
	int *new_cands = (int *)palloc(sizeof(int)*subst_len);
	int cands_num = 0;
	int new_cands_num = 0;
	int this_clen;
	char *this_cand = NULL;
	int has_match = -1;
	int i,j;
	int cand_step = 0;
	match_ent_t *matches = NULL;
	int matches_count = 0;
	int new_len = 0, new_curr = 0, curr = 0;
	char buff[100];

	for(i=0; i < subst_len; i++)
	{
		candidats[i] = i;
	}
	cands_num = subst_len;

	for(i=0; i < len; i++)
	{
		for(j=0; j < cands_num; j++)
		{
			this_cand = subst_array[candidats[j]];
			this_clen = strlen(this_cand);
			if(this_clen > cand_step)
			{
				if(this_cand[cand_step] == str[i])
				{
					if(cand_step ==  this_clen - 1)
					{
						has_match = candidats[j];
					}
					else
					{
						new_cands[new_cands_num++] = candidats[j];
					}
				}
			}
		}
		if(has_match >= 0)
		{
			for(j=0; j < subst_len; j++)
			{
				candidats[j] = j;
			}
			matches_count++;
			matches = 
				matches ? 
					repalloc(matches, (sizeof(match_ent_t) * matches_count)):
					palloc(sizeof(match_ent_t) * matches_count);
			matches[matches_count-1].start = i - cand_step;
			matches[matches_count-1].end = i;

			matches[matches_count-1].subst_len = sprintf(buff, "%d", has_match + step);
			matches[matches_count-1].subst = palloc(sizeof(matches[matches_count-1].subst_len));
			memcpy(matches[matches_count-1].subst, buff, matches[matches_count-1].subst_len);

			cands_num = subst_len;
			new_cands_num = 0;
			cand_step = 0;
			has_match = -1;
		}
		else if(new_cands_num == 0)
		{
			i -= cand_step;
			cand_step = 0;
			for(j=0; j < subst_len; j++)
			{
				candidats[j] = j;
			}
			cands_num = subst_len;
			new_cands_num = 0;
		}
		else
		{
			cand_step++;
			cands_num = new_cands_num;
			for(j=0; j < new_cands_num; j++)
			{
				candidats[j] = new_cands[j];
			}
			new_cands_num = 0;
		}
	}
	pfree(candidats);
	pfree(new_cands);

	if(matches_count > 0)
	{
		for(i=0; i < matches_count; i++)
		{
			new_len += (matches[i].end - matches[i].start + 1) - matches[i].subst_len ;
		}
		new_len = len - new_len;
		new_str = palloc(sizeof(char) * new_len+1);

		for(i=0; i < matches_count; i++)
		{
			memcpy(&(new_str[new_curr]), &(str[curr]), matches[i].start - curr);

			new_curr += (matches[i].start - curr);
			memcpy(&new_str[new_curr], matches[i].subst, matches[i].subst_len);
			new_curr += matches[i].subst_len;
			curr = matches[i].end+1;

			pfree(matches[i].subst);
		}
		if(curr < (len-1))
		{
			memcpy(&new_str[new_curr], &(str[curr]), len - curr);
			new_curr += (len - curr);
		}
		new_str[new_curr] = 0;
		pfree(matches);
		pfree(str);
		return new_str;
	}

	return str;
}

void destroyCronEnt(cron_ent_t *e)
{
	if(e->mins) pfree(e->mins);
	if(e->hour) pfree(e->hour);
	if(e->day) pfree(e->day);
	if(e->month) pfree(e->month);
	if(e->wdays) pfree(e->wdays);

	pfree(e);
}

cron_ent_t *parse_crontab(char *cron_str)
{
	cron_ent_t *CE;
	/* min, hour, day, month, wd */
	char *entrs[5] = {NULL, NULL, NULL, NULL, NULL};
	int N = strlen(cron_str);
	char tmp[256];
	int i, ti =0, en = 0 ;
	
	if(N == 7 && strcmp(cron_str, "@reboot") == 0)
	{
		CE = (cron_ent_t *)palloc(sizeof(cron_ent_t));
		memset((void *)CE, 0, sizeof(cron_ent_t)); 
		CE->onstart = 1;

		return CE;
	}

	for(i=0; i < N; i++)
	{
		if(cron_str[i] != ' ' && cron_str[i] != '\t')
		{
			tmp[ti++] = cron_str[i];
		}
		else
		{
			tmp[ti++] = 0;
			entrs[en] = palloc(sizeof(char) * ti);
			memcpy(entrs[en], tmp, ti);
			ti = 0;
			en++;
		}
	}
	if(ti)
	{
		tmp[ti++] = 0;
		entrs[en] = palloc(sizeof(char) * ti);
		memcpy(entrs[en], tmp, ti);
	}
	if(en != 4)
	{
		_cps_set_error(55, "cron string wrong format");
		for(i=0; i < 5; i++)
		{
			if(entrs[i]) pfree(entrs[i]);
		}
		return NULL;
	}

	CE = (cron_ent_t *)palloc(sizeof(cron_ent_t));
	memset((void *)CE, 0, sizeof(cron_ent_t)); 


	entrs[3] = _cps_subst_str(entrs[3], cps_month_subst_data, 12, 1);
	entrs[4] = _cps_subst_str(entrs[4], cps_wday_subst_data, 7, 0); 

	CE->mins = _cps_parse_range(entrs[0], 0, 59, &(CE->mins_len));
	if(CE->mins) CE->hour = _cps_parse_range(entrs[1], 0, 23, &(CE->hour_len));
	if(CE->hour) CE->day = _cps_parse_range(entrs[2], 1, 31, &(CE->day_len));
	if(CE->day) CE->month = _cps_parse_range(entrs[3], 1, 12, &(CE->month_len));
	if(CE->month) CE->wdays = _cps_parse_range(entrs[4], 0, 6, &(CE->wdays_len));

	for(i=0; i < 5; i++)
	{
		if(entrs[i]) pfree(entrs[i]);
	}

	if(CE->wdays) return CE;

	destroyCronEnt(CE);
	return NULL;
}

char *parse_crontab_to_json_text(char *cron_str)
{
	char *out = NULL;
	char *tmp_out;
	cron_ent_t *cron;

	cron = parse_crontab(cron_str);

	if(cron)
	{
		out = _cps_append_string(out, "{ \"crontab\": \"");
		out = _cps_append_string(out, cron_str);
		out = _cps_append_string(out, "\", ");
		if(cron->onstart)
		{
			out = _cps_append_string(out, "\"onstart\": 1");
		}
		else
		{
			out = _cps_append_string(out, "\"minutes\": ");
			tmp_out = _cps_make_array(cron->mins, cron->mins_len);
			out = _cps_append_string(out, tmp_out);
			pfree(tmp_out);
			out = _cps_append_string(out, ", ");

			out = _cps_append_string(out, "\"hours\": ");
			tmp_out = _cps_make_array(cron->hour, cron->hour_len);
			out = _cps_append_string(out, tmp_out);
			pfree(tmp_out);
			out = _cps_append_string(out, ", ");

			out = _cps_append_string(out, "\"days\": ");
			tmp_out = _cps_make_array(cron->day, cron->day_len);
			out = _cps_append_string(out, tmp_out);
			pfree(tmp_out);
			out = _cps_append_string(out, ", ");

			out = _cps_append_string(out, "\"months\": ");
			tmp_out = _cps_make_array(cron->month, cron->month_len);
			out = _cps_append_string(out, tmp_out);
			pfree(tmp_out);
			out = _cps_append_string(out, ", ");

			out = _cps_append_string(out, "\"wdays\": ");
			tmp_out = _cps_make_array(cron->wdays, cron->wdays_len);
			out = _cps_append_string(out, tmp_out);
			pfree(tmp_out);
		}
		out = _cps_append_string(out, "}");

		destroyCronEnt(cron);
	}

	return out;
}

