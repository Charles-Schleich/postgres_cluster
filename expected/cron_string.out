create user __temp_robot;
create user __temp_root WITH  SUPERUSER;
SET SESSION AUTHORIZATION __temp_root;
select schedule.create_job(
    '{
        "name": "Test @reboot",
        "cron": "@reboot",
        "command": "show all",
        "run_as": "__temp_robot"
    }'
);
 create_job 
------------
          1
(1 row)

select schedule.create_job(
    '{
        "name": "Test 1",
        "cron": "* * * * *",
        "command": "select ''this is every minute job''",
        "run_as": "__temp_robot",
        "last_start_available": "2 hours"
    }'
);
 create_job 
------------
          2
(1 row)

select schedule.create_job(
    '{
        "name": "Test 2 4/4 2/4 * * *",
        "cron": "4/4 2/4 * * *",
        "command": "select pg_sleep(10)",
        "run_as": "__temp_robot"
    }'
);
 create_job 
------------
          3
(1 row)

select schedule.create_job(
	'{
		"name": "Test 3",
		"cron": "23 1 * * THU,SUN",
		"command": "select ''ok'' as ok"
	}'
);
 create_job 
------------
          4
(1 row)

select id, node, name, rule, do_sql, same_transaction, postpone, retry from schedule.cron order by id;
 id |  node  |         name         |                                                                                                                                                                                                                                                                                             rule                                                                                                                                                                                                                                                                                             |                do_sql                 | same_transaction | postpone  | retry 
----+--------+----------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------+------------------+-----------+-------
  1 | master | Test @reboot         | {"crontab": "@reboot", "onstart": 1}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | {"show all"}                          | f                |           |     0
  2 | master | Test 1               | {"days": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31], "hours": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23], "wdays": [0, 1, 2, 3, 4, 5, 6], "months": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12], "crontab": "* * * * *", "minutes": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59]} | {"select 'this is every minute job'"} | f                | @ 2 hours |     0
  3 | master | Test 2 4/4 2/4 * * * | {"days": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31], "hours": [2, 6, 10, 14, 18, 22], "wdays": [0, 1, 2, 3, 4, 5, 6], "months": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12], "crontab": "4/4 2/4 * * *", "minutes": [4, 8, 12, 16, 20, 24, 28, 32, 36, 40, 44, 48, 52, 56]}                                                                                                                                                                                                                                             | {"select pg_sleep(10)"}               | f                |           |     0
  4 | master | Test 3               | {"days": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31], "hours": [1], "wdays": [0, 4], "months": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12], "crontab": "23 1 * * THU,SUN", "minutes": [23]}                                                                                                                                                                                                                                                                                                                              | {"select 'ok' as ok"}                 | f                |           |     0
(4 rows)

RESET SESSION AUTHORIZATION;
drop user __temp_root;
drop user __temp_robot;
